kernel/src/mem/phys.rs

```diff
@@ -113,33 +114,33 @@ pub(in crate::mem) fn init_stage2() {
     2. USABLE regions are sorted by base address, lowest to highest
     3. USABLE regions are 4KiB aligned (address and length)
      */
-    let highest_usable_address = {
-        let last_usable_region = regions
-            .iter()
-            .rev()
-            .find(|r| r.entry_type == EntryType::USABLE)
-            .expect("no usable regions");
-        last_usable_region.base + last_usable_region.length
-    };
+    
+    // Build memory regions for usable regions
+    let mut memory_regions = Vec::new();
```

Comment:
You should preallocate the vec, since reallocating could result in physical memory fragmentation, which could cause leaks in stage1 (because that thing can't deallocate)